File: Iteration-2/src/Student.java
import java.time.Year;
import java.util.ArrayList;
import org.apache.log4j.Logger;

public class Student extends Person {
    private Logger log = Logger.getLogger(Student.class);
    private StudentId studentId;
    private Transcript transcript;
    private Advisor advisor;
    private ArrayList<Course> requestedCourses;
    private int registrationYear;
    private int order;
    private int currentYear;
    private int semester;
    private ArrayList<String> studentOutput;

    public Student(String firstName, String lastName, int registrationYear, int order) {
        super(firstName, lastName);
        this.registrationYear = registrationYear;
        this.order = order;
        this.requestedCourses = new ArrayList<Course>();
        this.currentYear = Year.now().getValue();
        this.transcript = new Transcript();
        this.studentId = new StudentId(registrationYear, order);
        this.studentOutput = new ArrayList<String>();
        log.info("Student created =>  " + "Name: " + this.getStudentName() + " RegistrationYear: "
                + this.getRegistrationYear() + " Order: " + this.getOrder());
    }

    // GETTER & SETTER
    public ArrayList<String> getStudentOutput() {
        return studentOutput;
    }

    public void setStudentOutput(ArrayList<String> studentOutput) {
        this.studentOutput = studentOutput;
        /* log.info(this.getStudentName() + ": student output changed."); */
    }

    public String getStudentName() {
        return getFirstName() + " " + getLastName();
    }

    public Transcript getTranscript() {
        return this.transcript;
    }

    public StudentId getStudentId() {
        return this.studentId;
    }

    public Advisor getAdvisor() {
        return this.advisor;
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
        /* log.info(this.getStudentName() + ": student advisor changed."); */
    }

    public ArrayList<Course> getRequestedCourses() {
        return this.requestedCourses;
    }

    public int getRegistrationYear() {
        return this.registrationYear;
    }

    public int getOrder() {
        return this.order;
    }

    public int getCurrentYear() {
        return this.currentYear;
    }

    public int getSemester() {
        return this.semester;
    }

    public void setSemester(int semester) {
        this.semester = semester;
        log.info(this.getStudentName() + ": student semester changed.");
    }
}

File: Iteration-3/Code/Student.py
import logging
from Transcript import Transcript
from StudentId import StudentId
from Person import Person
from Advisor import Advisor
from Internship import Internship
import datetime


class Student(Person):

    def __init__(self, firstName: str, lastName: str, registrationYear: int, order: int):
        super().__init__(firstName, lastName)
        self.__registrationYear = registrationYear
        self.__order = order
        self.__requestedCourses = []
        self.__currentYear = int(datetime.datetime.now().year)
        self.__transcript = Transcript()
        self.__studentId = StudentId(registrationYear, order)
        self.__studentOutput = []
        self.__semester: int
        self.__advisor: Advisor
        logging.info("Student created =>  " + "Name: " + self.getStudentName() + " RegistrationYear: " +
                     str(self.getRegistrationYear()) + " Order: " + str(self.getOrder()))

    def willBeMultipleInternshipsInRequestedCourses(self):
        for course in self.__requestedCourses:
            if isinstance(course, Internship):
                return False
        return True

    def getStudentOutput(self):
        return self.__studentOutput

    def setStudentOutput(self, studentOutput):
        self.__studentOutput = studentOutput

    def getStudentName(self):
        return self.getFirstName() + " " + self.getLastName()

    def getTranscript(self):
        return self.__transcript

    def getStudentId(self):
        return self.__studentId

    def getAdvisor(self) -> Advisor:
        return self.__advisor

    def setAdvisor(self, advisor):
        self.__advisor = advisor

    def getRequestedCourses(self):
        return self.__requestedCourses

    def getRegistrationYear(self):
        return self.__registrationYear

    def getOrder(self):
        return self.__order

    def getCurrentYear(self):
        return self.__currentYear

    def getSemester(self) -> int:
        return self.__semester

    def setSemester(self, semester):
        self.__semester = semester
        logging.info(self.getStudentName() + ": student semester changed as " +
                     str(self.getSemester()))
